Mathlib/SetTheory/Ordinal/Veblen.lean

```diff
@@ -0,0 +1,308 @@
+    veblenWith f a 0 = veblenWith f b 0 ↔ a = b :=
+  (veblenWith_zero_strictMono hf hp).injective.eq_iff
+
+theorem left_le_veblenWith (hf : IsNormal f) (hp : 0 < f 0) (a b : Ordinal) :
+    a ≤ veblenWith f a b :=
+  (veblenWith_zero_strictMono hf hp).le_apply.trans <|
+    veblenWith_right_monotone hf _ (Ordinal.zero_le _)
+
+theorem isNormal_veblenWith_zero (hf : IsNormal f) (hp : 0 < f 0) :
+    IsNormal (veblenWith f · 0) := by
+  rw [isNormal_iff_strictMono_limit]
+  refine ⟨veblenWith_zero_strictMono hf hp, fun o ho a IH ↦ ?_⟩
+  rw [veblenWith_of_ne_zero f ho.pos.ne', derivFamily_zero]
+  apply nfpFamily_le fun l ↦ ?_
+  suffices ∃ b < o, List.foldr _ 0 l ≤ veblenWith f b 0 by
+    obtain ⟨b, hb, hb'⟩ := this
+    exact hb'.trans (IH b hb)
+  induction l with
+  | nil => use 0; simp [ho.pos]
+  | cons a l IH =>
+    obtain ⟨b, hb, hb'⟩ := IH
+    refine ⟨_, ho.succ_lt (max_lt a.2 hb), (veblenWith_right_monotone hf _ <| hb'.trans <|
+      veblenWith_left_monotone hf _ <| (le_max_right a.1 b).trans (Order.le_succ _)).trans ?_⟩
+    rw [veblenWith_veblenWith_of_lt hf]
+    rw [Order.lt_succ_iff]
+    exact le_max_left _ b
+
+theorem veblenWith_lt_veblenWith_iff (hf : IsNormal f) {a b c d : Ordinal} :
+    veblenWith f a b < veblenWith f c d ↔
+    a = c ∧ b < d ∨ a < c ∧ b < veblenWith f c d ∨ c < a ∧ veblenWith f a b < d := by
+  obtain h | rfl | h := lt_trichotomy a c
+  · simp_rw [h, h.ne, h.not_lt, false_and, false_or, or_false, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_lt_veblenWith_right_iff hf]
+  · simp [veblenWith_lt_veblenWith_right_iff hf]
+  · simp_rw [h, h.ne', h.not_lt, false_and, false_or, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_lt_veblenWith_right_iff hf]
+
+theorem veblenWith_le_veblenWith_iff (hf : IsNormal f) {a b c d : Ordinal} :
+    veblenWith f a b ≤ veblenWith f c d ↔
+    a = c ∧ b ≤ d ∨ a < c ∧ b ≤ veblenWith f c d ∨ c < a ∧ veblenWith f a b ≤ d := by
+  obtain h | rfl | h := lt_trichotomy a c
+  · simp_rw [h, h.ne, h.not_lt, false_and, false_or, or_false, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_le_veblenWith_right_iff hf]
+  · simp [veblenWith_le_veblenWith_right_iff hf]
+  · simp_rw [h, h.ne', h.not_lt, false_and, false_or, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_le_veblenWith_right_iff hf]
+
+theorem veblenWith_eq_veblenWith_iff (hf : IsNormal f) {a b c d : Ordinal} :
+    veblenWith f a b = veblenWith f c d ↔
+    a = c ∧ b = d ∨ a < c ∧ b = veblenWith f c d ∨ c < a ∧ veblenWith f a b = d := by
+  obtain h | rfl | h := lt_trichotomy a c
+  · simp_rw [h, h.ne, h.not_lt, false_and, false_or, or_false, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_inj hf]
+  · simp [veblenWith_inj hf]
+  · simp_rw [h, h.ne', h.not_lt, false_and, false_or, true_and]
+    conv_lhs => rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_inj hf]
```

Comment:
Perhaps `hf : IsNormal f` could be a `variable` here (you'd have to use `include`, I think). This could also be in the `IsNormal` namespace to allow dot notation on `hf`

Reply:
I agree that having this variable makes the code cleaner. I feel like having `IsNormal.veblen` and `veblen`, or even `IsNormal.veblenWith` and `veblen`, would be somewhat confusing, though.

Reply:
Perhaps this should be a `section`, then?

Reply:
That's what I did :)

Reply:
I meant only for the `IsNormal` part, but I guess it doesn't matter